### DISCUSSION TITLE
Specifying all struct/class members at once

### DISCUSSION THREAD

**Discussion by gurbaaz27** [nlohmann/json#3507] [Q&A] [answered]
Hi! I am currently using `NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE_WITH_DEFAULT(name, member1, member2, ...)` macro for a project to avoid boilerplate of `from_json/to_json` functions, and as the documentation states, 
"In all macros, the first parameter is the name of the class/struct, and all remaining parameters name the members. "

I wanted to state all members with a single variable in the parameter inside macro, because I will be needing all member variables of my struct (around 15-20 count), and writing all those within the macro one by one (i.e. (name, member1, member2, member3,....,member19) and so on) becomes tedious.

Is there an shortcut to do it? Some macro or even some variable within C++, or any other elegant way?

**Comment by nlohmann** [comment, marked as answer]
No, this is a limitation of C++.

**Reply by falbrechtskirchinger** [reply]
Maybe in C++26 or C++29. :crossed_fingers:

**Reply by gurbaaz27** [reply]
@nlohmann Thanks for the reply. I also had one more query if you (or anyone else) could take it up.

Is it possible for the library to automatically typecast the value from json when constructing a class/struct from json object. e.g.

`
struct foo {
   int a;
   string b;
  NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE_WITH_DEFAULT(foo, a, b);
};
 `, but the file `"input.json"` is

`
{
   "a": "100",
   "b": "bar"
}
`

Or is it NECESSARY that json file should contain `"a": 100` instead of `"a": "100"`? If yes, can you suggest any approach, since I would really be using macros `NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE_WITH_DEFAULT` instead of constructing `to_json/from_json` functions from scratch (as I have a rather complex json file).

**Reply by gurbaaz27** [reply]
Or is there any way to skip that element and move forward when I encounter a type error?

**Reply by falbrechtskirchinger** [reply]
> `{ "a": "100", "b": "bar" }`
> 
> Or is it NECESSARY that json file should contain `"a": 100` instead of `"a": "100"`? If yes, can you suggest any approach, since I would really be using macros `NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE_WITH_DEFAULT` instead of constructing `to_json/from_json` functions from scratch (as I have a rather complex json file).

Not possible using `NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE_WITH_DEFAULT`. You could define your own macro if you want to avoid writing `to_json/from_json` functions.

> Or is there any way to skip that element and move forward when I encounter a type error?

No. Same as above,
